Use vector for dtors in thread and remove defer()

jeehoonkang opened the pull request:
I replaced a custom list with `Vec` and corrected comments.

Fixes the point 1 of #29 
Fixes the point 3 of #29 
Closes #33 (actually using vector for dtors was intertwined with removing defer(); sorry for making a duplicated PR.)

cc @oliver-giersch @stjepang

jeehoonkang commented:
Before merging it, I'd like to add tests first.  While I'm quite confident it's correct, but my feeling is often wrong... :)

oliver-giersch commented:
For consistency, the vector should probably be reverse iterated during `drop_all()`, since the linked list uses a push_front/pop_front combination.

stjepang commented:
What should the following do?

```rust
catch_unwind(AssertUnwindSafe(|| {
    crossbeam::scope(|scope| {
        scope.defer(|| panic!());
        panic!();
    });
})
```

I'd say "catch the second panic and execute the deferred function", but this will currently abort the process. The reason is that the deferred function is executed while we're already unwinding (inside `Scope::drop`), which leads to abort.

To solve this problem, I suggest removing the `Drop` impl for `Scope` and executing the scope function like this:

```rust
// Executes the scoped function.
let result = catch_unwind(AssertUnwindSafe(|| f(&scope)));
scope.drop_all();
let ret = result?;
```

Can we add the example above as a test to make sure double panics don't abort the process?

> I think deferred functions should be executed before joining threads.

I agree, but... What is going to happen once we allow borrowing `Scope` inside scoped threads? Also, what if we allow  borrowing `Scope` inside deferred functions? In that case, deferred functions can spawn new threads and spawned threads can defer new functions. Then we'd have to make the order unspecified and just keep joining/executing until there are no more threads/functions.

jeehoonkang commented:
@stjepang I agree with you: there should be no order between thread joins and deferred functions, and `f(&scope)` should be `catch_unwind`ed.

After thinking of this issue a little bit more, I finally agree with you in removing `defer()` at all (#33).  First of all, it became clear that we still don't have an agreed semantics of `defer()` because of subtle issues discussed here and #33.  Second, I think whatever we want for `defer()` can be implemented on top of `scope()` without `defer()`.  As discussed in the context of logging (#26),  we can declare a container of deferred function before starting a `scope()`, add deferred functions to the container when necessary, and then execute it after `scope()`.

Maybe we can provide this deferred function container and logger as a companion to the `thread` library (related #27, cc @oliver-giersch).  What do you think?

jeehoonkang commented on src/thread.rs:
```diff
@@ -254,9 +240,13 @@ impl<'a> Scope<'a> {
     fn drop_all(&mut self) -> thread::Result<()> {
         let mut ret = Ok(());
-        while let Some(dtor) = DtorChain::pop(&mut self.dtors.borrow_mut()) {
+        let dtors = mem::replace(&mut *self.dtors.borrow_mut(), Vec::new());
```
Now this PR uses `drain(..)` :)  Thanks for suggestion!

jeehoonkang commented on src/thread.rs:
```diff
@@ -152,25 +152,11 @@ where
 pub struct Scope<'a> {
     /// The list of the deferred functions and thread join jobs.
-    dtors: RefCell<Option<DtorChain<'a, thread::Result<()>>>>,
+    dtors: RefCell<Vec<Box<FnBox<thread::Result<()>> + 'a>>>,
```
I'd like to switch to `dyn Trait` when we're starting to support Rust 2018.  Probably we have other lint warnings for Rust 2018, and I want to remove them at once.  What do you think?

jeehoonkang commented:
@stjepang @oliver-giersch I updated this PR as follows:

- I just removed `defer()`.  In a near future, I intend to make a PR that implements `defer()` *on top of* scoped threads.
- I added the test @stjepang mentioned (`panic_twice()`).

I think it's ready to be merged.  Would you please give another round of reviews?  Thanks!

stjepang commented on src/thread.rs:
```diff
@@ -210,39 +196,49 @@ pub struct ScopedJoinHandle<'scope, T: 'scope> {
+
+    // Joins all the threads.
+    scope.drop_all();
+    let panic = scope.panics.borrow_mut().pop();
```
Before this line, we could do:

```rust
let result = match result {
    Ok(r) => r,
    Err(e) => panic::resume_unwind(e),
};
```

And return the result as `Ok(result)` at the end of function.

This way we only catch panics inside spawned threads, not inside the scope itself.

Do you agree with this, or do you think we should catch panics inside the scope, too? I'm personally not 100% sure yet...

stjepang replied:
Yeah, following whatever Rayon does makes sense.

I'll also ask one more person for opinion, `prz` on IRC, who's been following recent developments in `crossbeam-utils`.

stjepang replied:
So there are a number of differences between scopes in crossbeam and rayon:

* In crossbeam, spawned threads can return a result. In rayon, they can't.
* In crossbeam, spawned threads can be manually joined. In rayon, they can't.
* In crossbeam, panics from spawned threads can be handles using `.join()`. In rayon, they can't.
* In crossbeam, panics from unjoined threads can be handled as the `Result` returned from `scope`. In rayon, `scope` simply propagates all panics and returns a `T`, not `Result`.

Basically, it looks like in rayon all panics are propagated, while in crossbeam panics tend to be caught and handled manually.

jeehoonkang replied:
I agree -- let's return `Result` instead of propagating panic.  I'm merging it.  Thanks!
